bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AbstractComponentImpl.java

```diff
@@ -50,4 +73,44 @@ public String getId() {
     public String getExportedType() {
         return resource.getResourceType();
     }
+
+    /**
+     * Returns an auto generated component ID.
+     *
+     * The ID is an SHA hash of the component path, prefixed with the component name.
+     *
+     * If the component is referenced, the path is taken to be a concatenation of the component path,
+     * with the path of the first parent context resource that exists on the page or in the template.
+     * This ensures the ID is unique if the same component is referenced multiple times
+     * on the same page or template.
+     *
+     * @return the auto generated component ID
+     */
+    private String generateId() {
+        String resourceType = resource.getResourceType();
+        String prefix = StringUtils.substringAfterLast(resourceType, "/");
+        String path = resource.getPath();
+        PageManager pageManager = currentPage.getPageManager();
+        Page containingPage = pageManager.getContainingPage(resource);
+        Template template = currentPage.getTemplate();
+        Boolean inCurrentPage = (containingPage != null && StringUtils.equals(containingPage.getPath(), currentPage.getPath()));
+        Boolean inTemplate = (template != null && path.startsWith(template.getPath()));
+        if (!inCurrentPage && !inTemplate) {
+            ComponentContext parentContext = componentContext.getParent();
+            while (parentContext != null) {
+                Resource parentContextResource = parentContext.getResource();
+                if (parentContextResource != null) {
+                    Page parentContextPage = pageManager.getContainingPage(parentContextResource);
+                    inCurrentPage = (parentContextPage != null && StringUtils.equals(parentContextPage.getPath(), currentPage.getPath()));
+                    inTemplate = (template != null && parentContextResource.getPath().startsWith(template.getPath()));
+                    if (inCurrentPage || inTemplate) {
+                        path = parentContextResource.getPath().concat(resource.getPath());
+                        break;
+                    }
+                }
+                parentContext = parentContext.getParent();
+            }
+        }
+        return prefix + "-" + StringUtils.substring(DigestUtils.shaHex(path), 0, 10);
```

Comment:
don't we generate the same ID when two the same XF will be placed on the same page?

Reply:
@lewandowskidawid - no we don't, we check for components referenced from other pages (including experience fragment pages) and in that case we hash a combination of the path of the context i.e. the referencing component and the path of the referenced component.